SIL.Core/IO/DirectoryHelper.cs

```diff
@@ -106,21 +108,45 @@ public static bool IsEmpty(string path, bool onlyCheckForFiles = false)
 
 		/// <summary>
 		/// Return subdirectories of <paramref name="path"/> that are not system or hidden.
+		/// Subdirectories which the user does not have permission to access will also be skipped.
 		/// There are some cases where our call to Directory.GetDirectories() throws.
 		/// For example, when the access permissions on a folder are set so that it can't be read.
 		/// Another possible example may be Windows Backup files, which apparently look like directories.
 		/// </summary>
 		/// <param name="path">Directory path to look in.</param>
-		/// <returns>Zero or more directory names that are not system or hidden.</returns>
-		/// <exception cref="UnauthorizedAccessException">E.g. when the user does not have
-		/// read permission.</exception>
+		/// <returns>Zero or more directory names that are not system or hidden</returns>
+		/// <exception cref="UnauthorizedAccessException">The caller does not have the required
+		/// permission to access the subdirectories of <paramref name="path"/>.</exception>
+		/// <exception cref="ArgumentException">
+		/// <paramref name="path" /> is a zero-length string, contains only white space, or contains one or more invalid characters. You can query for invalid characters by using the <see cref="M:System.IO.Path.GetInvalidPathChars" /> method.</exception>
+		/// <exception cref="ArgumentNullException">
+		/// <paramref name="path" /> is <see langword="null" />.</exception>
+		/// <exception cref="PathTooLongException">The specified <paramref name="path" />, file name,
+		/// or both exceed the system-defined maximum length.</exception>
+		/// <exception cref="IOException"><paramref name="path" /> is a file name.</exception>
+		/// <exception cref="DirectoryNotFoundException">The specified <paramref name="path" /> is
+		/// invalid (for example, it is on an unmapped drive).</exception>
 		public static string[] GetSafeDirectories(string path)
 		{
-			return (from directoryName in Directory.GetDirectories(path)
-				let dirInfo = new DirectoryInfo(directoryName)
-				where (dirInfo.Attributes & FileAttributes.System) != FileAttributes.System
-				where (dirInfo.Attributes & FileAttributes.Hidden) != FileAttributes.Hidden
-				select directoryName).ToArray();
+			var list = new List<string>();
+			foreach (var directoryName in Directory.GetDirectories(path))
+			{
+				try
+				{
+					var dirInfo = new DirectoryInfo(directoryName);
+					if ((dirInfo.Attributes & FileAttributes.System) != FileAttributes.System)
+					{
+						if ((dirInfo.Attributes & FileAttributes.Hidden) != FileAttributes.Hidden)
+							list.Add(directoryName);
+					}
+				}
+				catch (SecurityException e)
+				{
+					Console.WriteLine(e);
```

Comment:
I wonder if it would be better to use `Debug.WriteLine` or `Trace.WriteLine`.

Reply:
FWIW: There are 156 other instances of Console.Write(Line) in libpalaso and only a tiny handful of Trace.WriteLine calls. Maybe that makes sense, but it might bear a review. In this case, since it is not really a debug-only issue but something that could happen for real end-users, I don't think Debug.WriteLine is appropriate. If we hit this situation and the ende result is something undesirable (either a subsequent crash or failure or something the user reports as a potential bug), I think we would want the log file to reflect the error, so I'm thinking of changing these to Logger.WriteError instead.